Map submission `subjects` to Metax `field of science`

csc-jm opened the pull request:
### Description

<!-- Please include a summary of the change or any information deemed important. -->
PR adds functionality to map `subjects` from submission doi info to Metax `field_of_science`. As discussed in #327 and further described in #506, the idea is to revert from using the OECD classification for subject terms and instead use the YSO fields of science. Hence there are some changes to `submission.json`.

### Related issues
<!-- Reference any follow up issues with "Fixes #<issue-num>." -->
Fixes #506 

### Type of change

<!-- Replace [ ] with [x] to select options. -->
<!-- Please delete options that are not relevant. -->

- [x] New feature

### Changes Made

<!-- List changes made. -->
- Added a new metax mapping method
- Added new values to the metax reference fetching script
- Made the metax reference file to be a static file in this repository
- Updated the `submission.json` schema and `specification.yml` of docs
- Methods in `submission.py` add more subject info to `doiInfo`
- Added some tests
- Fixed some spelling errors

### Testing

<!-- Are any tests part of this PR. -->
<!-- Replace [ ] with [x] to select options. -->
<!-- Please delete options that are not relevant. -->

- [x] Unit Tests
- [x] Integration Tests

### Mentions
<!-- Shout outs to your friends that you made this happen or need help. -->

csc-felipe commented on scripts/metax_mappings/create_metax_references.py:
```diff
@@ -69,5 +69,15 @@
 
+
+# parsing codes and identifiers for mapping fields of science
+fos_codes_url = "https://metax.demo.fairdata.fi/es/reference_data/field_of_science/_search?size=100"
```
Are you getting all of them when `size=100`?

csc-jm replied:
There are currently 74 different FOS in the metax reference data list so that query `size` gets them all. Not sure if they'll ever add more

csc-felipe commented:
Please check with @hannyle if it affects the frontend. And if necessary, create a ticket there.

csc-jm commented on metadata_backend/helpers/schemas/submission.json:
```diff
@@ -148,67 +148,114 @@
+                                    "3125 - Otorhinolaryngology, ophthalmology",
+                                    "3126 - Surgery, anesthesiology, intensive care, radiology",
+                                    "313 - Dentistry",
+                                    "3141 - Health care science",
+                                    "3142 - Public health care science, environmental and occupational health",
+                                    "315 - Sport and fitness sciences",
+                                    "316 - Nursing",
+                                    "317 - Pharmacy",
+                                    "318 - Medical biotechnology",
+                                    "319 - Forensic science and other medical sciences",
+                                    "4 - Agriculture and forestry",
+                                    "4111 - Agronomy",
+                                    "4112 - Forestry",
+                                    "412 - Animal science, dairy science",
+                                    "413 - Veterinary science",
+                                    "414 - Agricultural biotechnology",
+                                    "415 - Other agricultural sciences",
+                                    "5 - Social sciences",
+                                    "511 - Economics",
+                                    "512 - Business and management",
+                                    "513 - Law",
+                                    "5141 - Sociology",
+                                    "5142 - Social policy",
+                                    "515 - Psychology",
+                                    "516 - Educational sciences",
+                                    "517 - Political science",
+                                    "518 - Media and communications",
+                                    "519 - Social and economic geography",
+                                    "520 - Other social sciences",
+                                    "6 - Humanities",
+                                    "611 - Philosophy",
+                                    "6121 - Languages",
+                                    "6122 - Literature studies",
+                                    "6131 - Theatre, dance, music, other performing arts",
+                                    "6132 - Visual arts and design",
+                                    "614 - Theology",
+                                    "615 - History and archaeology",
+                                    "616 - Other humanities",
+                                    "9 - Other",
+                                    "999 - Other"
```
If the frontend utilizes this list of values for creating submissions, then this shouldn't break anything

blankdots replied:
it was hardcoded as FOS https://github.com/CSCfi/metadata-submitter-frontend/blob/master/src/features/wizardSubmissionFolderSlice.tsx#L227-L228

blankdots replied:
we can fix that in a later iteration of front-end, and probably we can alleviate that by making some changes in the backend

blankdots replied:
it is good at is, we can address changes when front-end starts to work on it

csc-felipe commented:
Please, add an entry to the changelog.

csc-felipe reviewed (changes requested):
Looks great!

I'd still like to see an entry in the changelog, and it would be better if the generated file would be included in the repo as part of a separate PR.

csc-jm commented:
Changelog updated and the generated file was removed from this PR. However the unit tests I wrote for this PR fail now, but I swear they work if you generate the references :D 

(hence this could be merged and the unit tests will pass soon, once we add the generated file back)

csc-jm commented:
Ok so I cleaned up this PR with the magic of interactive rebasing. I also commented out the unit tests that are included here because they don't work without the static reference files which are added in #568. PR #568 is based on this PR so the unit tests written here pass in that PR.

csc-felipe commented:
Nice! It's a bit funny how you made the tests in this PR, but run them only later, but I can live with it :)

csc-jm commented on metadata_backend/services/metax_service_handler.py:
```diff
@@ -224,7 +225,13 @@ async def update_dataset_with_doi_info(self, datacite_info: Dict, _metax_ids: Li
-            mapped_metax_data = mapper.map_metadata()
+            try:
+                mapped_metax_data = mapper.map_metadata()
+            except SubjectNotFoundException as error:
```
Yea I originally didn't have this because I too thought that the subjects should always correspond with the schema. But then I figured you can technically still feed the method faulty values which would cause issues. Will it ever happen? Maybe not, but if it ever does, this avoids returning a 500 response.

csc-felipe replied:
Yes, makes sense.

blankdots reviewed (changes requested):
this fixes metax but I think it breaks a bit integration with datacite as the FOS is something recommended by datacite along with https://id.loc.gov/authorities/subjects.html

This requires changes to docs `specification.yml` but i marked it as a request for change until we change the change viable for datacite as well

csc-felipe commented:
> this fixes metax but I think it breaks a bit integration with datacite as the FOS is something recommended by datacite along with https://id.loc.gov/authorities/subjects.html

Could you elaborate on the link? I didn't see how is it connected to the subjects we are using.

According to Datacite specification on the metadata schema, the `subject` field is a free text field, with the possibility to indicate the scheme and URI of the resource we used. It seems pretty clear that we don't need to be limited to using FOS or LCSH. I am no specialist in research and what are valid subject classifications, but it does seem to me that we do have an alternative, and perhaps more detail could be added to the PR description about it.
![image](https://user-images.githubusercontent.com/93575941/189596236-aaab5815-b720-418f-a934-832d7caac1b6.png)
Page 18-19 of https://schema.datacite.org/meta/kernel-4/doc/DataCite-MetadataKernel_v4.4.pdf

csc-felipe commented:
Or then what we should do, is manually map from Datacite to Metax. It's not so many, so it's quite doable.

blankdots reviewed (changes requested):
i managed to come to terms with using something else than FOS, after checking it is ok, however still some changes required:

i think the `test_doi.json`, `subjects` should contain only:
```
  "subjects": [
    {
      "subject": "999 - Other",
    }
  ]
  ```
  
  and the rest:
```
      "subjectScheme": "Korkeakoulujen tutkimustiedonkeruussa käytettävä tieteenalaluokitus",
      "schemeUri": "http://www.yso.fi/onto/okm-tieteenala/conceptscheme",
      "valueUri": "http://www.yso.fi/onto/okm-tieteenala/ta999",
      "classificationCode": "999"
```

is added from the code in submissions.py as part of `_prepare_datacite_study` and `_prepare_datacite_dataset`

csc-jm commented on metadata_backend/helpers/schemas/submission.json:
```diff
@@ -148,67 +148,114 @@
                     "description": "Subject, classification code, or key phrase describing the resources specified by OECD Fields of Science and Technology (FOS)",
```
Changed

csc-jm commented:
Updated the `specification.yml` with the new subject names/codes, fixed the descriptions in the `submission.json` schema, moved the subject info to be assigned in `_prepare_datacite_study` and `_prepare_datacite_dataset` and rebased with `develop` branch.
